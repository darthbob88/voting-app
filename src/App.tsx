import React from 'react';
import './App.css';
import CastBallot from './components/CastBallot/CastBallot';
import NewPoll from './components/NewPoll/NewPoll';
import PollListing from './components/PollListing/PollListing';
import PollResults from './components/PollResults/PollResults';

function App() {

  // TODO: Fix this before actually deploying it.
  localStorage.setItem("best-fruit", JSON.stringify({
    "ID": "best-fruit",
    "name": "Best Fruit",
    "creator": "darthbob88@gmail.com",
    "creation": 1668465233437,
    "expiration": 1668551633437,
    "method": "FPTP",
    "candidates": [
      "apple",
      "orange",
      "banana",
      "mango"
    ],
    "ballots": [
      {
        "ballotID": "butts",
        "votes": "apple",
        "timestamp": 1670282960239,
        "pollID": "best-fruit",
        "voterID": "butts"
      },
      {
        "ballotID": "butts",
        "voterID": "butts",
        "votes": "orange",
        "timestamp": 1671854055125,
        "pollID": "best-fruit"
      },
      {
        "ballotID": "butts",
        "voterID": "butts",
        "votes": "mango",
        "timestamp": 1671854347738,
        "pollID": "best-fruit"
      },
      {
        "ballotID": "butts",
        "voterID": "butts",
        "votes": "mango",
        "timestamp": 1671854350413,
        "pollID": "best-fruit"
      },
      {
        "ballotID": "butts",
        "voterID": "butts",
        "votes": "orange",
        "timestamp": 1671854352933,
        "pollID": "best-fruit"
      },
      {
        "ballotID": "butts",
        "voterID": "butts",
        "votes": "mango",
        "timestamp": 1671854367541,
        "pollID": "best-fruit"
      }
    ]
  }));
  localStorage.setItem("best-pet", JSON.stringify({
    "ID": "best-pet",
    "name": "Best Pet",
    "creator": "darthbob88@gmail.com",
    "creation": 1703207900702,
    "expiration": 1703294300702,
    "method": "IRV",
    "candidates": [
      "dog",
      "cat",
      "goldfish",
      "hamster"
    ],
    "ballots": []
  }))

  return (
    <div className="App">
      <header className="App-header">
        <PollListing />
        <NewPoll />
        <CastBallot pollID='best-fruit' />
        <PollResults pollID='best-fruit' />
      </header>
    </div>
  );
}

export default App;
