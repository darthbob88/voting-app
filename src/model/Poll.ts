export type Poll = {
    ID: string,
    name: string,
    creator: string,
    creation: number,
    expiration: number,
    method: string,
    candidates: string[],
    ballots: Ballot[]
}

export type Ballot = {
    ballotID: string,
    voterID: string,
    pollID: string,
    votes: any,
}