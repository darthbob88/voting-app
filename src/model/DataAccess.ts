import { Ballot, Poll } from "./Poll";

/*
 * Using local storage, just so I don't have to set up
 * Firebase until I'm more confident in the data structures. 
 */
//TODO: Set up Firebase, or whatever other data storage I decide to use.

export const createPoll = (newPoll: Poll) => {
    if (localStorage.getItem(newPoll.ID) != null) {
        throw new Error(`Invalid poll ${newPoll.ID}`);
    }
    localStorage.setItem(newPoll.ID, JSON.stringify(newPoll));
    return newPoll.ID;
}

export const retrievePoll = (pollID: string): Poll | null => {
    const jsonBlob = localStorage.getItem(pollID);
    if (jsonBlob == null) {
        throw new Error(`Invalid poll ${pollID}`);
    }

    return JSON.parse(jsonBlob);

}

export const retrievePollList = (): string[] => {
    return Object.keys(localStorage);
}

export const castBallot = (newBallot: Ballot, pollID: string) => {
    const poll = retrievePoll(pollID);
    if (poll == null) {
        throw new Error(`Invalid poll ${pollID}`);
    }
    poll.ballots.push(newBallot);
    localStorage.setItem(poll.ID, JSON.stringify(poll));
}