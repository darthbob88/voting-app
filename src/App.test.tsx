import { render } from '@testing-library/react';
import App from './App';
import { createPoll } from './model/DataAccess';
import { testPoll } from './model/Poll';

test('renders home page stuff', () => {
  createPoll({ ...testPoll, ID: "best-fruit" });
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});
