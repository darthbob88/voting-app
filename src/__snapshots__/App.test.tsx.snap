// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders home page stuff 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <div>
        <h2>
          List of Polls
        </h2>
        <h4>
          Does not currently work
        </h4>
        <ul>
          <li>
            <a
              href="best-fruit"
            >
              best-fruit
            </a>
             
          </li>
        </ul>
      </div>
      <form>
        <h3>
          New Poll Form
        </h3>
        <label>
          <input
            type="text"
            value="pollName"
          />
           Please enter a name for your poll
        </label>
        <label
          class="label"
        >
          <input
            type="text"
            value="apple,orange,banana"
          />
           Please enter a comma-separated list of candidates for your poll
        </label>
        <label>
          <select
            data-testid="methodSelect"
          >
            <option
              data-testid="methodOption"
              value="FPTP"
            >
              First past the post
            </option>
            <option
              data-testid="methodOption"
              value="IRV"
            >
              Instant runoff vote
            </option>
             
          </select>
           Please select a method for your poll
        </label>
        <button>
          Create poll
        </button>
      </form>
      <form>
        <h3>
          Ballot for 
          pollName
        </h3>
        <div
          class="ballot"
        >
          <label>
             
            <input
              name="ballot"
              type="radio"
              value="apple"
            />
            apple
          </label>
          <label>
             
            <input
              name="ballot"
              type="radio"
              value="orange"
            />
            orange
          </label>
          <label>
             
            <input
              name="ballot"
              type="radio"
              value="banana"
            />
            banana
          </label>
          <label>
             
            <input
              name="ballot"
              type="radio"
              value="mango"
            />
            mango
          </label>
        </div>
        <div>
          Current value of ballot: 
        </div>
        <div>
          Poll method is 
          FPTP
          , options are 
          ["apple","orange","banana","mango"]
        </div>
        <button>
          Cast Ballot
        </button>
      </form>
      <div>
        <h3>
          Results for 
          pollName
        </h3>
        <div
          class="ballot"
        >
          Winner is 
          mango
           with 
          3
           votes out of 
          6
          <br />
          Detailed result:
          <ol>
            <li>
              mango
               - 
              3
               votes
            </li>
            <li>
              orange
               - 
              2
               votes
            </li>
            <li>
              apple
               - 
              1
               votes
            </li>
            <li>
              banana
               - 
              0
               votes
            </li>
          </ol>
        </div>
      </div>
    </header>
  </div>
</div>
`;
